File: vlc-android/src/org/videolan/vlc/gui/preferences/PreferencesExtensions.java
package org.videolan.vlc.gui.preferences;

import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import com.google.android.material.appbar.AppBarLayout;
import androidx.preference.Preference;
import androidx.preference.PreferenceCategory;
import androidx.preference.PreferenceScreen;
import androidx.appcompat.widget.SwitchCompat;
import android.view.View;

import org.videolan.vlc.R;
import org.videolan.vlc.extensions.ExtensionListing;
import org.videolan.vlc.extensions.ExtensionsManager;
import org.videolan.vlc.gui.view.ClickableSwitchPreference;
import org.videolan.vlc.util.Settings;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class PreferencesExtensions extends BasePreferenceFragment {

    private List<ExtensionListing> mExtensions = new ArrayList<>();
    private SharedPreferences mSettings;
    private PreferenceScreen preferenceScreen;
    private int count = 0;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mSettings = Settings.INSTANCE.getInstance(requireContext());
        mExtensions = ExtensionsManager.getInstance().getExtensions(getActivity().getApplication(), false);
        preferenceScreen = this.getPreferenceScreen();
    }

    @Override
    public void onStart() {
        super.onStart();
        ((AppBarLayout) ((PreferencesActivity)getActivity()).findViewById(R.id.appbar)).setExpanded(true, false);
        createCheckboxes();
    }

    @Override
    public void onStop() {
        super.onStop();
        preferenceScreen.removeAll();
    }

    @Override
    protected int getXml() {
        return R.xml.preferences_extensions;
    }

    @Override
    protected int getTitleId() {
        return R.string.extensions_prefs_category;
    }

    private void createCheckboxes() {
        PackageManager pm = getActivity().getApplicationContext().getPackageManager();
        for (int i = 0 ; i < mExtensions.size() ; ++i) {
            ExtensionListing extension = mExtensions.get(i);
            ClickableSwitchPreference switchPreference = new ClickableSwitchPreference(preferenceScreen.getContext());
            switchPreference.setTitle(extension.title());
            switchPreference.setSummary(extension.description());
            final String key = ExtensionsManager.EXTENSION_PREFIX + "_" + extension.componentName().getPackageName();
            switchPreference.setKey(key);
            int iconRes = extension.menuIcon();
            Drawable extensionIcon = null;
            if (iconRes != 0) {
                try {
                    Resources res = pm.getResourcesForApplication(extension.componentName().getPackageName());
                    extensionIcon = res.getDrawable(extension.menuIcon());
                } catch (PackageManager.NameNotFoundException e) {}
            }
            if (extensionIcon != null)
                switchPreference.setIcon(extensionIcon);
            else
                try {
                    switchPreference.setIcon(pm.getApplicationIcon(mExtensions.get(i).componentName().getPackageName()));
                } catch (PackageManager.NameNotFoundException e) {
                    switchPreference.setIcon(R.drawable.icon);
                }
            final boolean checked = mSettings.getBoolean(key, false);
            switchPreference.setChecked(checked);
            preferenceScreen.addPreference(switchPreference);
            switchPreference.setOnSwitchClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (((SwitchCompat)view).isChecked())
                        mSettings.edit().putBoolean(key, true).apply();
                    else
                        for (Map.Entry<String, ?> entry : mSettings.getAll().entrySet())
                            if (entry.getKey().startsWith(ExtensionsManager.EXTENSION_PREFIX + "_"))
                                mSettings.edit().putBoolean(entry.getKey(), false).apply();
                }
            });
            count++;
        }

        if (count == 0) {
            PreferenceCategory emptyCategory = new PreferenceCategory(preferenceScreen.getContext());
            emptyCategory.setTitle(R.string.extensions_empty);
            preferenceScreen.addPreference(emptyCategory);
        }
    }

    @Override
    public boolean onPreferenceTreeClick(Preference preference) {
        String key = preference.getKey();
        if (key == null || !key.startsWith(ExtensionsManager.EXTENSION_PREFIX + "_"))
            return false;
        PreferencesExtensionFragment fragment = new PreferencesExtensionFragment();
        Bundle extras = new Bundle();
        extras.putString("extension_key", key);
        fragment.setArguments(extras);
        loadFragment(fragment);
        return super.onPreferenceTreeClick(preference);
    }
}

File: application/vlc-android/src/org/videolan/vlc/gui/preferences/PreferencesExtensions.kt
package org.videolan.vlc.gui.preferences

import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.View
import androidx.appcompat.widget.SwitchCompat
import androidx.core.content.ContextCompat
import androidx.core.os.bundleOf
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import com.google.android.material.appbar.AppBarLayout
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.ObsoleteCoroutinesApi
import org.videolan.vlc.R
import org.videolan.vlc.extensions.ExtensionListing
import org.videolan.vlc.extensions.ExtensionsManager
import org.videolan.vlc.gui.view.ClickableSwitchPreference
import org.videolan.tools.Settings
import org.videolan.tools.putSingle
import java.util.*

@ObsoleteCoroutinesApi
@ExperimentalCoroutinesApi
class PreferencesExtensions : BasePreferenceFragment() {

    private var extensions: List<ExtensionListing> = ArrayList()
    private lateinit var settings: SharedPreferences
    private var count = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        settings = Settings.getInstance(requireContext())
        extensions = ExtensionsManager.getInstance().getExtensions(requireActivity().applicationContext, false)
        preferenceScreen = this.preferenceScreen
    }

    override fun onStart() {
        super.onStart()
        ((activity as PreferencesActivity).findViewById<View>(R.id.appbar) as AppBarLayout).setExpanded(true, false)
        createCheckboxes()
    }

    override fun onStop() {
        super.onStop()
        preferenceScreen!!.removeAll()
    }

    override fun getXml(): Int {
        return R.xml.preferences_extensions
    }

    override fun getTitleId(): Int {
        return R.string.extensions_prefs_category
    }

    private fun createCheckboxes() {
        val pm = requireActivity().applicationContext.packageManager
        for (i in extensions.indices) {
            val extension = extensions[i]
            val switchPreference = ClickableSwitchPreference(preferenceScreen!!.context)
            switchPreference.title = extension.title()
            switchPreference.summary = extension.description()
            val key = ExtensionsManager.EXTENSION_PREFIX + "_" + extension.componentName().packageName
            switchPreference.key = key
            val iconRes = extension.menuIcon()
            var extensionIcon: Drawable? = null
            if (iconRes != 0) {
                try {
                    extensionIcon = ContextCompat.getDrawable(requireActivity(), extension.menuIcon())
                } catch (e: PackageManager.NameNotFoundException) {
                }

            }
            if (extensionIcon != null)
                switchPreference.icon = extensionIcon
            else
                try {
                    switchPreference.icon = pm.getApplicationIcon(extensions[i].componentName().packageName)
                } catch (e: PackageManager.NameNotFoundException) {
                    switchPreference.setIcon(R.drawable.icon)
                }

            val checked = settings.getBoolean(key, false)
            switchPreference.isChecked = checked
            preferenceScreen!!.addPreference(switchPreference)
            switchPreference.setOnSwitchClickListener(View.OnClickListener {
                if ((view as SwitchCompat).isChecked)
                    settings.putSingle(key, true)
                else
                    for ((key1) in settings.all)
                        if (key1.startsWith(ExtensionsManager.EXTENSION_PREFIX + "_"))
                            settings.putSingle(key1, false)
            })
            count++
        }

        if (count == 0) {
            val emptyCategory = PreferenceCategory(preferenceScreen!!.context)
            emptyCategory.setTitle(R.string.extensions_empty)
            preferenceScreen!!.addPreference(emptyCategory)
        }
    }

    override fun onPreferenceTreeClick(preference: Preference): Boolean {
        val key = preference.key
        if (key == null || !key.startsWith(ExtensionsManager.EXTENSION_PREFIX + "_"))
            return false
        val fragment = PreferencesExtensionFragment()
        fragment.arguments = bundleOf("extension_key" to key)
        loadFragment(fragment)
        return super.onPreferenceTreeClick(preference)
    }
}
